### PR TITLE
Rudimentary progress tracking.

### PR DESCRIPTION
**Pull request by ryzhyk**
See individual commits.

### CODE REVIEW

File: crates/adapters/src/controller/mod.rs [hunk 1]
```diff
@@ -228,6 +230,26 @@ impl Controller {
         Ok(())
     }
 
+    /// Check whether the pipeline has processed all input data to completion.
+    ///
+    /// This method should only be used with pipelines whose input endpoints
```
Comment (mbudiu-vmw) [inline crates/adapters/src/controller/mod.rs:235]: what happens if it's not?
won't it just return 'false'?

File: crates/adapters/src/controller/mod.rs [hunk 2]
```diff
@@ -228,6 +230,26 @@ impl Controller {
...
+    /// signalling the end of input, e.g., files.  It returns `true` when
+    /// the following conditions are satisfied.
+    ///
+    /// * All input endpoints have signalled end-of-input.
+    /// * All input records received from all endpoints have been processed by
+    ///   the circuit.
+    /// * All output records have been sent to respective output transport
+    ///   endponts.
+    ///
+    /// Note that, depending on the type and configuration of the output
```
Comment (mbudiu-vmw) [inline crates/adapters/src/controller/mod.rs:246]: Is this note really necessary?

File: crates/adapters/src/controller/mod.rs [hunk 3]
```diff
@@ -428,8 +463,11 @@ impl InputEndpointDescr {
 }
 
 /// A lock-free queue used to send output batches from the circuit thread
-/// to output endpoint threads.
-type BatchQueue = SegQueue<Vec<Box<dyn SerBatch>>>;
+/// to output endpoint threads.  Each entry is annotated with a progress label
+/// that is equal to the number of input records fully processed by
+/// DBSP before emitting this batch of outputs.  The label increases
+/// monotonically over time.
+type BatchQueue = SegQueue<(Vec<Box<dyn SerBatch>>, u64)>;
```
Comment (mbudiu-vmw) [inline crates/adapters/src/controller/mod.rs:470]: yes another kind of virtual time

File: demo/demo_notebooks/fraud_detection.ipynb [hunk 4]
```diff
@@ -58,7 +58,7 @@
     "test_path            = \"/home/leonid/projects/database-stream-processor/demo/demo_notebooks/fraud_data/test.csv\"\n",
     "test_outpath         = \"/home/leonid/projects/database-stream-processor/demo/demo_notebooks/fraud_data/test_output.csv\"\n",
     "demographics_path    = \"/home/leonid/projects/database-stream-processor/demo/demo_notebooks/fraud_data/demographics.csv\"\n",
-    "simulation_path      = \"/home/leonid/projects/database-stream-processor/demo/demo_notebooks/fraud_data/simulation.csv\"\n",
+    "simulation_path      = \"/home/leonid/projects/database-stream-processor/demo/demo_notebooks/fraud_data/simulation_short.csv\"\n",
```
Comment (mbudiu-vmw) [inline demo/demo_notebooks/fraud_detection.ipynb:61]: are we supposed to review notebooks?
I'd like to, but this is not really readable

File: crates/adapters/src/controller/stats.rs [hunk 5]
```diff
@@ -423,6 +518,15 @@ impl InputEndpointStatus {
             .fetch_add(num_records, Ordering::AcqRel)
     }
 
+    fn eoi(&self, num_records: u64) {
```
Comment (mbudiu-vmw) [inline crates/adapters/src/controller/stats.rs:521]: set_eoi?

**Review by mbudiu-vmw — APPROVED** [review]

**Review by Kixiron — APPROVED** [review]

File: crates/adapters/src/controller/mod.rs (hunk 2 above)
Comment (ryzhyk) [reply crates/adapters/src/controller/mod.rs:246]: you're right

File: demo/demo_notebooks/fraud_detection.ipynb (hunk 4 above)
Comment (ryzhyk) [reply demo/demo_notebooks/fraud_detection.ipynb:61]: It's ok to ignore them for now.

File: crates/adapters/src/controller/mod.rs (hunk 1 above)
Comment (ryzhyk) [reply crates/adapters/src/controller/mod.rs:235]: you're right

File: crates/adapters/src/controller/mod.rs (hunk 2 above)
Comment (ryzhyk) [reply crates/adapters/src/controller/mod.rs:246]: I think it's useful.

**Pull request merged**